File: python/spider_browser/ai/observe.py
"""observe() — element discovery (works WITHOUT LLM)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from ..protocol.protocol_adapter import ProtocolAdapter
from ..utils.dom import GET_INTERACTIVE_ELEMENTS
from ..utils.html import truncate_html
from .llm_provider import LLMProvider


@dataclass
class ObserveResult:
    """An observed interactive element on the page."""

    selector: str
    tag: str
    type: str
    text: str
    aria_label: str
    placeholder: str
    href: str
    value: str
    rect: Dict[str, float]
    score: Optional[float] = None


async def observe(
    adapter: ProtocolAdapter,
    instruction: Optional[str] = None,
    llm: Optional[LLMProvider] = None,
) -> List[ObserveResult]:
    """
    Discover interactive elements on the page.

    Works WITHOUT an LLM — injects a DOM traversal script to collect
    interactive elements with selectors, text, and bounding rects.

    When instruction + LLM are provided, adds ranking/filtering.
    """
    raw = await adapter.evaluate(GET_INTERACTIVE_ELEMENTS)

    if not isinstance(raw, list) or len(raw) == 0:
        return []

    elements = [
        ObserveResult(
            selector=el.get("selector", ""),
            tag=el.get("tag", ""),
            type=el.get("type", ""),
            text=el.get("text", ""),
            aria_label=el.get("ariaLabel", ""),
            placeholder=el.get("placeholder", ""),
            href=el.get("href", ""),
            value=el.get("value", ""),
            rect=el.get("rect", {}),
        )
        for el in raw
        if isinstance(el, dict)
    ]

    if not instruction or not llm:
        return elements

    # Use LLM to rank/filter by relevance
    lines = []
    for i, el in enumerate(elements):
        parts = [f"[{i}] <{el.tag}>"]
        if el.text:
            parts.append(f'text="{el.text}"')
        if el.aria_label:
            parts.append(f'aria="{el.aria_label}"')
        if el.placeholder:
            parts.append(f'placeholder="{el.placeholder}"')
        if el.href:
            parts.append(f'href="{el.href}"')
        if el.type:
            parts.append(f'type="{el.type}"')
        lines.append(" ".join(parts))

    element_summary = "\n".join(lines)

    response = await llm.chat_json([
        {
            "role": "system",
            "content": (
                'You are an element selector. Given a list of page elements and an instruction, '
                'return a JSON object with an "indices" array of element indices that match the instruction. '
                'Order by relevance (most relevant first). Return {"indices": []} if none match.'
            ),
        },
        {
            "role": "user",
            "content": f"Instruction: {instruction}\n\nElements:\n{element_summary}",
        },
    ])

    indices = response.get("indices", []) if isinstance(response, dict) else []
    result = []
    for rank, idx in enumerate(indices):
        if isinstance(idx, int) and 0 <= idx < len(elements):
            el = elements[idx]
            el.score = 1 - rank / max(len(indices), 1)
            result.append(el)
    return result

File: go/observe.go
package spiderbrowser

import (
	"encoding/json"
	"fmt"
	"strings"
)

// ObserveResult represents an observed interactive element on the page.
type ObserveResult struct {
	Selector    string `json:"selector"`
	Tag         string `json:"tag"`
	Type        string `json:"type"`
	Text        string `json:"text"`
	AriaLabel   string `json:"ariaLabel"`
	Placeholder string `json:"placeholder"`
	Href        string `json:"href"`
	Value       string `json:"value"`
	Rect        struct {
		X      int `json:"x"`
		Y      int `json:"y"`
		Width  int `json:"width"`
		Height int `json:"height"`
	} `json:"rect"`
	Score float64 `json:"score,omitempty"`
}

// Observe discovers interactive elements on the page.
// Works without an LLM. When instruction is provided + LLM is available, adds ranking.
func Observe(adapter *ProtocolAdapter, instruction string, llm LLMProvider) ([]ObserveResult, error) {
	val, err := adapter.Evaluate(getInteractiveElementsJS)
	if err != nil {
		return nil, err
	}

	// Convert the result to ObserveResult slice
	var elements []ObserveResult
	data, err := json.Marshal(val)
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal(data, &elements); err != nil {
		return nil, nil // No elements found
	}

	if len(elements) == 0 {
		return nil, nil
	}

	// If no instruction or no LLM, return all elements
	if instruction == "" || llm == nil {
		return elements, nil
	}

	// Use LLM to rank/filter
	var summary strings.Builder
	for i, el := range elements {
		fmt.Fprintf(&summary, "[%d] <%s>", i, el.Tag)
		if el.Text != "" {
			fmt.Fprintf(&summary, " text=%q", el.Text)
		}
		if el.AriaLabel != "" {
			fmt.Fprintf(&summary, " aria=%q", el.AriaLabel)
		}
		if el.Placeholder != "" {
			fmt.Fprintf(&summary, " placeholder=%q", el.Placeholder)
		}
		if el.Href != "" {
			fmt.Fprintf(&summary, " href=%q", el.Href)
		}
		if el.Type != "" {
			fmt.Fprintf(&summary, " type=%q", el.Type)
		}
		summary.WriteString("\n")
	}

	var result struct {
		Indices []int `json:"indices"`
	}
	err = llm.ChatJSON([]LLMMessage{
		{
			Role: "system",
			Content: "You are an element selector. Given a list of page elements and an instruction, " +
				"return a JSON object with an \"indices\" array of element indices that match the instruction. " +
				"Order by relevance (most relevant first). Return {\"indices\": []} if none match.",
		},
		{
			Role:    "user",
			Content: fmt.Sprintf("Instruction: %s\n\nElements:\n%s", instruction, summary.String()),
		},
	}, &result)
	if err != nil {
		return elements, nil // Fall back to all elements on LLM error
	}

	filtered := make([]ObserveResult, 0, len(result.Indices))
	for rank, idx := range result.Indices {
		if idx >= 0 && idx < len(elements) {
			el := elements[idx]
			el.Score = 1.0 - float64(rank)/float64(max(len(result.Indices), 1))
			filtered = append(filtered, el)
		}
	}
	return filtered, nil
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
